include/pybind11/pybind11.h

```diff
@@ -613,6 +613,39 @@ class cpp_function : public function {
                     call.args_convert.swap(second_pass_convert);
                 }
 
+
+                //printf("calling %s %s, args %d %lu\n", func.name, func.signature, func.nargs, func.args.size());
+                for (size_t i = 0; i < func.args.size(); ++i) {
+                    if (!func.args[i].disown) {
+                        //printf("NOT disowning arg %lu\n", i);
+                        continue;
+                    } else {
+                        //printf("DISOWNING arg %lu\n", i);
+                    }
+
+                    auto &arg = call.args[i];
+                    if (!arg)
+                        pybind11_fail("nothing to disown?!");
+
+                    if (arg.is_none())
+                        continue; /* Nothing to disown */
+
+                    auto inst = reinterpret_cast<detail::instance *>(arg.ptr());
+                    auto value_and_holder = values_and_holders(inst).begin();
+
+
+                    using holder_type = std::unique_ptr<void*>; // how do I get the proper one?
```

Comment:
@torokati44 After seeing your stuff, and reworking my PR, I believe this function should be what you need for type-erased holder transfer:
https://github.com/EricCousineau-TRI/pybind11/blob/379afa73c7fedded57e4fedc9c3b077cb7292643/include/pybind11/cast.h#L482

You could use it to something like the tune of:
```
if (!reclaim_existing_if_needed(inst, tinfo, existing_holder)) {
  throw cast_error("Cannot release ownership of this object");  // If you have an incompatible holder type, like `shared_ptr`.
}
```

The caveat is, you'll need to figure out what the holder type is for the given argument. I believe that should be elsewhere in this area of code.
(Also, ensure that you get `tinfo` from the argument type and NOT the holder, as `holder->type` may be null!)